### PR TITLE
Add UKAEA TGLF surrogate

### PR DESCRIPTION
**Pull request by theo-brown**
New PR for TGLF-NN-UKAEA (replacing #11).

This PR contains a JAX implementation of the model in [Zanisi et al 2024](https://iopscience.iop.org/article/10.1088/1741-4326/ad240d).

We are in the final stages of open-sourcing the model weights and test data under the LGPLv2.1. It is expected that these will be stored at [ukaea/tglfnn-ukaea](https://github.com/ukaea/tglfnn-ukaea), while architecture implementations will be stored elsewhere. `fusion_surrogates` is the natural home for the JAX version of this network, which is intended for use in TORAX.

Once the weights and test data are open, I will update the tests etc as needed.

### CODE REVIEW

File: fusion_surrogates/tglfnn_ukaea/config.py [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    normalize: bool = True
+    unnormalize: bool = True
+    hidden_size: int = 512
+
+    @classmethod
+    def load(cls, config_path: str):
+        with open(config_path, "r") as f:
+            config = yaml.safe_load(f)
+
+        return cls(
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply fusion_surrogates/tglfnn_ukaea/config.py:56]: Good catch. 

* I've added hidden size (although in all the networks we've developed it's the same).

* normalize=True and unnormalize=True are baked in to the architecture of the original models and so do not appear in the saved config.yaml. I still thought it best to expose them as a property of the Config object, because then all the potential settings are available.
   
* In the same vein, I've added activation as a property, with default 'relu', rather than hard-coding it in the model constructor.

File: fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_model_test.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    )
+
+    return model
+
+
+class TGLFNNukaeaModelTest(parameterized.TestCase):
+
+    @parameterized.named_parameters(
+        dict(
+            testcase_name="batched_inputs",
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_model_test.py:71]: Changed test logic and added more cases.

**Review by hamelphi — COMMENTED** [review]
LGTM, just a few more details to fix.

File: fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_config.py [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+
+MEAN_OUTPUT_IDX: Final[int] = 0
+VAR_OUTPUT_IDX: Final[int] = 1
+
+OutputLabel = Literal["efe_gb", "efi_gb", "pfi_gb"]
+OUTPUT_LABELS = typing.get_args(OutputLabel)
+
+Machine = Literal["step", "multimachine"]
+
+INPUT_LABELS_DICT: Final[Mapping[Machine, list[str]]] = {
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_config.py]: Switched for an enum and reran pytype, I think it's passing now.

**Comment by theo-brown** [comment]
Changes:
* Revert `import tglfnn_ukaea_config as config` to avoid overwriting
* Fix type hinting on all functions, including switching Literal to Enum for machine type (checked w/ pytype)
* Rename install option
* Add standalone test files (required adding conftest.py to parse abseil flags)

**Review by hamelphi — COMMENTED** [review]
Thanks for bearing with me with all the comments. Just a few more nits.

File: fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_model_test.py [hunk 4]
```diff
@@ -14,60 +14,56 @@
...
-        hidden_size=hidden_size,
-        machine="multimachine",
-    )
-    dummy_stats = config.TGLFNNukaeaModelStats(
-        input_mean=jnp.zeros(len(dummy_config.input_labels)),
-        input_std=jnp.zeros(len(dummy_config.input_labels)),
-        output_mean=jnp.zeros(len(dummy_config.output_labels)),
-        output_std=jnp.zeros(len(dummy_config.output_labels)),
-    )
+from fusion_surrogates.tglfnn_ukaea import tglfnn_ukaea_config, tglfnn_ukaea_model
```
[the remark this answers is not in the document]
Comment (theo-brown) [reply fusion_surrogates/tglfnn_ukaea/tglfnn_ukaea_model_test.py]: Sorry, this was an overactive autoformatter. Reverted here and elsewhere.

**Review by hamelphi — APPROVED** [review]
LGTM. I will retry the merge.

**Comment by hamelphi** [comment]
FYI. The merge seems to be working internally. I am just waiting for an internal LGTM to be able to merge back to the repo.
